File: problems/0090-subsets-ii/python/solution.py
def subsetsWithDup(nums):
    """Subsets II — https://leetcode.com/problems/subsets-ii/"""
    nums.sort()
    res = []
    subset = []

    def backtrack(start):
        res.append(subset[:])
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i - 1]:
                continue  # skip duplicate at this level
            subset.append(nums[i])
            backtrack(i + 1)
            subset.pop()

    backtrack(0)
    return res

File: problems/0090-subsets-ii/typescript/solution.ts
// Subsets II — https://leetcode.com/problems/subsets-ii/
export function subsetsWithDup(nums: number[]): number[][] {
  const sorted = [...nums].sort((a, b) => a - b);
  const res: number[][] = [];
  const subset: number[] = [];
  const backtrack = (start: number): void => {
    res.push([...subset]);
    for (let i = start; i < sorted.length; i++) {
      if (i > start && sorted[i] === sorted[i - 1]) continue; // skip duplicate at this level
      subset.push(sorted[i]);
      backtrack(i + 1);
      subset.pop();
    }
  };
  backtrack(0);
  return res;
}
